homeassistant/components/hydrawise/__init__.py

```diff
@@ -76,6 +93,53 @@ def hub_refresh(event_time):
     return True
 
 
+@dataclass
+class HydrawiseData:
+    """Container for data returned by our DataUpdateCoordinator."""
+
+    controllers: dict[int, Controller] = field(default_factory=dict)
+    """Controllers indexed by their id."""
+
+    zones: dict[int, dict[int, Zone]] = field(default_factory=dict)
+    """Zones indexed by their controller, then by their id."""
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Hydrawise from a config entry."""
+    hass.data.setdefault(DOMAIN, {})
+
+    auth = Auth(entry.data[CONF_USERNAME], entry.data[CONF_PASSWORD])
+    api = Pydrawise(auth)
+
+    async def async_update_data() -> HydrawiseData:
+        data = HydrawiseData()
+        for ctrl in await api.get_controllers():
+            data.controllers[ctrl.id] = ctrl
+            data.zones[ctrl.id] = {z.id: z for z in await ctrl.get_zones()}
+        return data
+
+    coordinator = DataUpdateCoordinator(
```

Comment:
Using a DataUpdateCoordinator is a great idea, but it should be either a preliminary PR or a follow-up PR.

Reply:
I appreciate the sentiment, but I'm not sure I agree? The config flow implementation here is essentially greenfield, so setting up the scaffolding using best-practices (and such that it won't send spurious API calls) seems like the correct thing to do.

If you feel strongly about this I guess I can back out just this part, but that seems like unnecessary work on my part.

Reply:
I hadn't realised that it was used only for the new implementation.
I suggest that you turn this into a `HydrawiseDataUpdateCoordinator` class, inside a new `coordinator.py` module and passing `api` as an argument.

You can then use `coordinator = HydrawiseDataUpdateCoordinator(api)`